### PR TITLE
[PPML] PPML overview draft

### PR DESCRIPTION
**Pull request by qiyuangong**
Addressing #3655

### CODE REVIEW

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 1]
```diff
@@ -1,14 +1,46 @@
-## PPML (Privacy Preserving Machine Learning)
-### PPML for Big Data AI
+# PPML (Privacy Preserving Machine Learning)
+
+Analtyics-Zoo PPML is an end to end framework for building privacy presrving maching leanrnig applications on top of large scale dataset.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:3]: end-to-end

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 2]
```diff
@@ -1,14 +1,46 @@
...
 - Scenario
 - Get started 
+
+```bash
+```
+
+
+## 
+
+## Reference
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:40]: Reference not needed

**Comment by jason-dai** [comment]
1) We should first explain what PPML is in the beginning of the document. Focusing on what its value is and what problem it solves for the

**Comment by jason-dai** [comment]
1) In the beginning, explain that what the PPML support in Analytics Zoo is - focusing on what problem it solves for the users using SGX. There is no need to explain why SGX is used and what different modules in Analytics Zoo are.

2) In "PPML for Big Data AI" section, first explain that it applies to the E2E Big Data AI pipeline, including big data analysis (such as Spark and Flink) and ML/DL; and then explain how it extends SGX TEE to a trusted cluster environment (including data encryption, TLS, remote attestation, etc.) for Big Data AI. 

3) In "Trusted Big Data Analytics and ML" section, first describe what features are supported (i.e., trusted Spark analytics, and BigDL based DL), and then provide user guide on how the user can use these features.

4) In "Trusted Realtime Compute and ML" section, first describe what features are supported (i.e., trusted Flink analytics, and Cluster Serving), and then provide user guide on how the user can use these features.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 3]
```diff
@@ -1,14 +1,196 @@
...
+
+To reslove this problem, Analytics-Zoo chooses [Intel SGX (Software Guard Extensions)](https://software.intel.com/content/www/us/en/develop/topics/software-guard-extensions.html), a widely used [TEE (Trusted Execution Environment)](https://en.wikipedia.org/wiki/Trusted_execution_environment) technology, as main security building block for this PPML platforms. Different from other PPML technologies, e.g., [HE (Homomorphic Encryption)](https://en.wikipedia.org/wiki/Homomorphic_encryption), [MPC (Multi-Party Computation) or SMC (Secure Multi-Party Computation)](https://en.wikipedia.org/wiki/Secure_multi-party_computation) and [DP (Differential Privacy)](https://en.wikipedia.org/wiki/Differential_privacy), Intel SGX performs well on all measures (security, performance and utility).
+
+<p align="center"><img src="../../../../../docs/Image/PPML/ppml_tech.png" height="180px"><br></p>
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:11]: This is not needed

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 4]
```diff
@@ -1,14 +1,196 @@
...
+
+Based on Intel SGX (Software Guard Extensions) and LibOS projects ([Graphene](https://grapheneproject.io/) and [Occlum](https://occlum.io/)), Analytics-Zoo empowers our customers (e.g., data scientists and big data developers) to build PPML applications on top of large scale dataset without impacting existing applications.
+
+![PPML Architecture](../../../../../docs/Image/PPML/ppml_arch.png#center)
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:15]: This is not needed

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 5]
```diff
@@ -1,14 +1,196 @@
...
+Note: Intel SGX requires hardware support, please [check if your CPU has this feature](https://www.intel.com/content/www/us/en/support/articles/000028173/processors/intel-core-processors.html). In [3rd Gen Intel Xeon Scalable Processors](https://newsroom.intel.com/press-kits/3rd-gen-intel-xeon-scalable/), SGX allows up to 1TB of data to be included in secure enclaves.
+
+### Key features
+
+- Protecting data and model confidentiality
+- Seamless migrate existing big data applications into privacy preserving applications
+- Trusted big data & AI Platform based on Intel SGX
+  - Trusted Big Data Analytics and ML
+  - Trusted Realtime Compute and ML
+  - Trusted Federated Learning (incoming)
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:26]: This is not needed

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 6]
```diff
@@ -1,14 +1,196 @@
...
+- Batch computation/analytics, i.e., privacy preserved Spark jobs
+- Interactive computation/analytics, i.e., privacy preserved SparkSQL
+- Large scale Spark related workload, e.g., TPC-H Benchmark
+- Distributed machine learning & deep Learning with BigDL
+
+### Get started
+
+#### Setp 0: Prepare Environment
+
+Please clone or download Analytics-Zoo source code, then enter `analytics-zoo/ppml`.
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:48]: we should not ask user to clone our source code

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 7]
```diff
@@ -1,14 +1,196 @@
...
+
+If SGX driver is not installed, please install SGX driver with this command
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:50]: how does the user know if SGX driver is installed? this step is for general SGX, not PPML, yes? maybe we should make it a prerequisite

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 8]
```diff
@@ -1,14 +1,196 @@
...
+./scripts/install-graphene-driver.sh
+```
+
+Prepare the keys for TLS.
+
+```bash
+./scripts/generate-keys.sh
+```
+
+Build docker image from Dockerfile
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:62]: does the user have to build from dockerfile?

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 9]
```diff
@@ -1,14 +1,196 @@
...
+cp -r ../../keys .
+./build-docker-image.sh
+```
+
+#### Step 1: Start Spark in SGX
+
+Enter `analytics-zoo/ppml/trusted-big-data-ml/scala/docker-graphene` dir. Start Spark service with this command
+
+```bash
+./start-local-big-data-ml.sh
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:75]: The user need to update `start-local-big-data-ml.sh` first, no? and we should allow the user to specify the name, instead using a fixed `spark-local`

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 10]
```diff
@@ -1,14 +1,196 @@
...
+sudo docker exec -it spark-local bash
+```
+
+##### Step 2: Submit jobs to Spark
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:79]: This is actually confusing; it actually tries to run the Spark Pi example to verify the environment, yes? Then you should describe its purpose, and maybe make it a subsection of step 1.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 11]
```diff
@@ -1,14 +1,196 @@
...
+
+Modify `init.sh`, which is task/env related.
+
+```bash
+./init.sh
+vim start-spark-local-pi-sgx.sh
+```
```
Comment (jason-dai) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:86]: we can just provide a script to run the Pi example for the user to test.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md (hunk 8 above)
Comment (glorysdj) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md:62]: there are two choices, 
1. pull our docker image from the docker hub, 
2. build the docker image by themselves.
we need to publish the docker images first.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md (hunk 9 above)
Comment (glorysdj) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md:75]: the names should be modified. This script only starts a container and next user run spark local[*] job on graphene in this container.

**Comment by jason-dai** [comment]
The "Trusted Big Data Analysis and ML" section can be organized a follows:

1) Use cases - what use cases it supports on the trusted platform
    - Big Data analysis using Spark (Spark SQL, Dataframe, MLlib, etc.)
    - Distributed deep learning using BigDL
   
2) Getting started 
    - Prerequisite: install SGX, download scripts, etc.
    - Step 0: prepare environment (build/pull docker, generate key, etc.)
    - Step 1: start the platform (stat container, verify spark, etc.)
    - Step 2: run Spark program (how to run arbitrary spark program, using Spark SQL as an example)
    - Step 3: run BigDL program (how to run arbitrary BigDL program, using LeNet as an example)

The "Trusted Realtime Compute and ML" section can be similar (showing Flink and Cluster Serving examples) @glorysdj @qiyuangong

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 12]
```diff
@@ -1,14 +1,213 @@
...
+
+Check SGX driver with `ls /dev | grep sgx`. If SGX driver is not installed, run `install-graphene-driver.sh` and run it with root premission.
+
+```bash
+./ppml/scripts/install-graphene-driver.sh
+```
+
+#### Setp 0: Prepare Environment
+
+Prepare keys for TLS.
```
Comment (qiuxin2012) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:60]: The generate-keys need a lot of typing, and I think we need to tell user the excepted result of this script. Yu wei's machine happened to restart, there is only a `server.key`, He didn't know what to do next.

Comment (qiyuangong) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md:60]: OK. WIll add key list and input warning.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 13]
```diff
@@ -1,14 +1,213 @@
...
+
+```bash
+./ppml/scripts/generate-keys.sh
```
Comment (qiuxin2012) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:63]: I think maybe we can provide some default keys for user.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 14]
```diff
@@ -1,14 +1,247 @@
...
-### Trusted Big Data Analytics and ML
-- Scenario
-- Get started 
- 
-### Trusted Realtime Compute and ML
-- Scenario
-- Get started 
+# PPML (Privacy Preserving Machine Learning)
+
+PPML (Privacy-Preserving Machine Learning) aims at protecting user privacy, meanwhile keep machine learning applications still useful. However, achieving this goal without impacting existing applications is difficult, especially in end-to-end big data scenarios. To reslove this problem, Analytics-Zoo provides an end-to-end PPML platform for Big Data AI based on Intel SGX (Software Guard Extensions). This PPML platform ensure the whole Big Data & AI pipeline are fully protected by secured SGX enclave in hardware level, meanwhile exising Big Data & AI applications, such as Flink, Spark, SparkSQL and machine/deep learning, can be seamlessly migrated into this PPML platform without any code changes.
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:3]: found some grammatic typos:
“meanwhile **keep** machine learning” ----》“meanwhile **keeping** machine learning”
“To **reslove** this problem” ----》 “To **resolve** this problem”
“This PPML platform **ensure**” ----》“This PPML platform **ensures** ”
“meanwhile **exising** Big Data” ----》“meanwhile **existing** Big Data”

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 15]
```diff
@@ -1,14 +1,247 @@
...
+
+## PPML for Big Data AI
+
+To take full advantage of big data, especially the value of private or sensitive data, customers need to build a trusted platform under the guidance of privacy laws or regulation, such as [GDPR](https://gdpr-info.eu/). This requirement raises big challenges to customers who already have big data and big data applications, such as Spark/SparkSQL, Flink/FlinkSQL and AI applications. Migrating these applications into privacy preserving way requires lots of additional efforts.
+
+With Analytics-Zoo, customers can build a Trusted Platform for big data with a few clicks, and all existing big data & AI applications, such as Flink and Spark applications, can be migrated into this platform without any code changes. In specific, Analytics-Zoo uses serval security technologies
+
+- Confiditional Computation with Intel SGX. Intel SGX provides hardware-based isolation and memory encryption with very limited attack surface.
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:11]: Some typos:
“**Confiditional** Computation” ----》"**Confidential** Computation"

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 16]
```diff
@@ -1,14 +1,247 @@
...
+- Seamless migration with LibOS. Based on LibOS projects ([Graphene](https://grapheneproject.io/) and [Occlum](https://occlum.io/)), Analytics-Zoo empowers our customers (e.g., data scientists and big data developers) to build PPML applications on top of large scale dataset without impacting existing applications.
+- Secured networks with TLS and encryption. All network traffic are protected by TLS, in some cases, content should be encrypted before transformation.
+- File or model protection with encryption. Model and senstive configuration files will be encrypted before uploading to Trusted platform. These files are only decrypted in SGX enclave.
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:14]: Typo:
"Model and **senstive** configuration files" ----> "Model and **sensitive** configuration files"

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 17]
```diff
@@ -1,14 +1,247 @@
...
+- Environment & App attestation with SGX attestation. SGX attestation ensure that remote/local SGX env and applications can be verified.
+
+
+Note: Intel SGX requires hardware support, please [check if your CPU has this feature](https://www.intel.com/content/www/us/en/support/articles/000028173/processors/intel-core-processors.html). In [3rd Gen Intel Xeon Scalable Processors](https://newsroom.intel.com/press-kits/3rd-gen-intel-xeon-scalable/), SGX allows up to 1TB of data to be included in secure enclaves.
+
+### Key features
+
+- Protecting data and model confidentiality
+  - Sensitive input/output data (computation, training and inference), e.g., healthcare data
+  - Propretary model, e.g., model trained with self-owned or sensitive data
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:24]: Typo:
"**Propretary** model, " ----> "**Proprietary** model"

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 18]
```diff
@@ -1,14 +1,247 @@
...
+### Use cases
+
+- Big Data analysis using Spark (Spark SQL, Dataframe, MLlib, etc.)
+- Distributed deep learning using BigDL
+
+### Get started
+
+#### Prerequisite: Install SGX Driver & Prepare Scripts
+
+Please check if current platform [has SGX feature](https://www.intel.com/content/www/us/en/support/articles/000028173/processors/intel-core-processors.html). Then, enable SGX feature in BIOS. Note that after SGX is enabled, a protion of memory will be assigned to SGX, and cannot be used or seen by OS.
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:43]: You can only put hyperlink for “SGX feature”
"a **protion** of memory " ----> "a **portion** of memory "

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 19]
```diff
@@ -1,14 +1,247 @@
...
+
+Check SGX driver with `ls /dev | grep sgx`. If SGX driver is not installed, please install [SGX DCAP driver](https://github.com/intel/SGXDataCenterAttestationPrimitives/tree/master/driver/linux) with [install-graphene-driver.sh](https://github.com/intel-analytics/analytics-zoo/blob/master/ppml/scripts/install-graphene-driver.sh) (need root premission).
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:45]: typo:
"need root **premission**" ----> "need root **permission**"

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 20]
```diff
@@ -1,14 +1,247 @@
...
+
+```bash
+./ppml/scripts/install-graphene-driver.sh
+```
+
+#### Setp 0: Prepare Environment
```
Comment (OpheliaLjh) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:51]: "**Setp** 0" ----> "**Step** 0"

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md (hunk 14 above)
Comment (qiyuangong) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md]: Tks :)
Addressed in new version.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 21]
```diff
@@ -1,14 +1,314 @@
...
+Also, you can build docker image from Dockerfile (this will take some time).
+
+```bash
+cd ppml/trusted-big-data-ml/scala/docker-graphene
+./build-docker-image.sh
+```
+
+#### Step 1: Single-Node Trusted Big Data Analytics and ML Platform
+
+Enter `analytics-zoo/ppml/trusted-big-data-ml/scala/docker-graphene` dir. Start Spark service with this command
```
Comment (glorysdj) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:103]: this is not needed? and conflict with below cd command

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md [hunk 22]
```diff
@@ -1,14 +1,314 @@
...
+
+Prepare `keys` and `password`
```
Comment (glorysdj) [inline docs/readthedocs/source/doc/PPML/Overview/ppml.md:105]: is this part needed? user can keep their keys anywhere they like, and only modify the ./start-local-big-data-ml.sh
we should ask user to modify ./start-local-big-data-ml.sh to specify some pathes and environments for the docker command

Comment (qiyuangong) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md:105]: OK. Will simplify or remove this part.

File: docs/readthedocs/source/doc/PPML/Overview/ppml.md (hunk 4 above)
Comment (qiyuangong) [reply docs/readthedocs/source/doc/PPML/Overview/ppml.md]: Removed images

**Pull request merged**